src/win/fs.c

```diff
@@ -2009,28 +2010,146 @@ static void fs__ftruncate(uv_fs_t* req) {
+
+  src_handle = CreateFileW(src,
+                           GENERIC_READ,
+                           FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                           NULL,
+                           OPEN_EXISTING,
+                           0,
+                           NULL);
+
+  if (src_handle == INVALID_HANDLE_VALUE) {
+    error = GetLastError();
+    goto exit;
+  }
+
+  dst_handle = CreateFileW(dst,
+                           GENERIC_WRITE,
+                           0,
+                           NULL,
+                           flags & UV_FS_COPYFILE_EXCL ? OPEN_EXISTING : 0,
+                           0,
+                           NULL);
+  if (dst_handle == INVALID_HANDLE_VALUE) {
+    error = GetLastError();
+    goto exit1;
+  }
+
+  if (fs__stat_handle(src_handle, &statbuf, 0) != 0) {
+    error = GetLastError();
+    goto exit2;
+  }
+
+  /* Truncate the file. Wish there was a ftruncate_handle. */
+  eof_info.EndOfFile.QuadPart = src_size = statbuf.st_size;
+  status = pNtSetInformationFile(dst_handle,
+                                 &io_status,
+                                 &eof_info,
+                                 sizeof eof_info,
+                                 FileEndOfFileInformation);
+
+  if (!NT_SUCCESS(status)) {
+    error = pRtlNtStatusToDosError(status);
+    goto exit2;
+  }
+
+  /* Do the cloning. Clones must fall by cluster boundary and may not be
+   * larger than 4GiB. */
+  for (; offset < src_size - GIG; offset += GIG) {
+    if (! fs__clonefile_dup_extents(src_handle, dst_handle, offset, GIG)) {
+      error = GetLastError();
+      goto exit2;
+    }
+  }
+
+  /* Try to clone the rest with cluster size. Size overflow is fine here. */
+  for (int i = 0; i < 2, i++) {
+    int64_t size = (src_size - offset + (CLUSTERSIZES[i] - 1)) / CLUSTERSIZES[i];
+    if (! fs__clonefile_dup_extents(src_handle, dst_handle, offset, size)) {
+      error = GetLastError();
+      goto exit2;
+    }
+    break;
+  }
+
+exit2:
+  CloseHandle(dst_handle);
+exit1:
+  CloseHandle(src_handle);
+exit:
+  return error;
+}
+
 
 static void fs__copyfile(uv_fs_t* req) {
   int flags;
   int overwrite;
   uv_stat_t statbuf;
   uv_stat_t new_statbuf;
-
+  DWORD error;
   flags = req->fs.info.file_flags;
 
-  if (flags & UV_FS_COPYFILE_FICLONE_FORCE) {
-    SET_REQ_UV_ERROR(req, UV_ENOSYS, ERROR_NOT_SUPPORTED);
-    return;
-  }
-
   overwrite = flags & UV_FS_COPYFILE_EXCL;
 
+  if (flags & UV_FS_COPYFILE_FICLONE ||
+      flags & UV_FS_COPYFILE_FICLONE_FORCE) {
+    error = fs__clonefile(req->file.pathw, req->fs.info.new_pathw, flags);
+    if (error) {
+      if (req->flags & UV_FS_COPYFILE_FICLONE_FORCE) {
+        SET_REQ_WIN32_ERROR(req, error);
+        goto err;
+      }
+      /* If it's not EEXIST we probably hit a copy error. Removing is hard, so
+         we do this. (If it's some other error it won't hurt to hit it again.) */
+      if (error != ERROR_ALREADY_EXISTS) {
```

Comment:
I'm not sure that we should ever turn on overwriting without the user knowing.

Reply:
Maybe I should signal the error source instead of just GetLastError in the return value. This way it's more clear whether it was a failed block copy or some other more fatal error.

I was doing this to avoid the unlink/delete/remove because calling it feels kind of verbose. The thinking was that we did create something here.

Reply:
It'd be better to track that explicitly, I think, perhaps by passing `&overwrite` as an argument.

Unrelated, for reference, here's the documentation for the most commonly encountered errors: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/6c268ef7-11c5-4830-ba0a-20fbe0557646